src/native/managed/cdacreader/Microsoft.Diagnostics.DataContractReader.Abstractions/Contracts/IReJIT.cs

```diff
@@ -2,13 +2,25 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System;
+using System.Collections.Generic;
 
 namespace Microsoft.Diagnostics.DataContractReader.Contracts;
 
+public enum RejitState
+{
+    Requested,
+    Active
+}
+
 internal interface IReJIT : IContract
 {
     static string IContract.Name { get; } = nameof(ReJIT);
+
     bool IsEnabled() => throw new NotImplementedException();
+
+    RejitState GetRejitState(ILCodeVersionHandle codeVersionHandle) => throw new NotImplementedException();
```

Comment:
With the cdac I feel like we should either make an effort to make the data structures and algorithms match the runtime directly or make a conscious effort to have a higher level API that abstracts away some of the details. Putting `GetRejitState` and `GetRejitId` here doesn't feel like it accomplishes either of those tasks. In the runtime they are part of `ILCodeVersion` and in the cdac they are now part of a different conceptual contract, but we don't change the abstraction level.

Curious to know what other people's thoughts are here.

Reply:
As far as I know, we do need contract methods to get this metadata (id/state) as the cDAC is setup to _not_ allow users to access the marshalled data structures directly. Enforcing the use of contract methods allows for versioning independent of runtime changes.

Another option would be to put these methods on the `ICodeVersions` contract. However, I believe separating the `ICodeVersions` contract from the `IReJIT` contract is beneficial as it allows us to separate the logic determining what code version is active from associated metadata which ReJIT/TieredCompilation add to code versions. In case this metadata structure changes in the future it will be simple to change just the relevant contract implementation without modifying the larger code versioning logic.

When TieredCompilation SOS APIs are implemented, I think we should create an `ITieredCompilation` contract that allows us to get the optimization tier metadata off of `NativeCodeVersions`.

Since `FEATURE_REJIT` can be enabled/disabled separately from `FEATURE_CODE_VERSIONING`, I think this also gives us a path to adapt to toggleable runtime features through contract versions. If for example ReJIT is disabled in a target, the ReJIT contract can be different and let cDAC users know ReJIT is disabled.

I am also curious if other people have suggestions on better ways to implement the ReJIT cDAC functionality.